### PR TITLE
"Save our build times" aka Add support for mounts during docker build

### LINKED ISSUE #14080: build time only -v option
As suggested by @cpuguy83 in https://github.com/docker/docker/issues/3156
here is the use case for a flexible -v option at build time.

When building a Docker image I need to install a database and an app. It's all wrapped up in two tarballs: 1 for the DB and 1 for the App that needs to be installed in it (schema, objects, static data, credentials etc.).  The whole solution is then run via a shell script that handles several shell variables and tune OS credentials and other things accordingly.
When I explode the above tarball (or use the Dockerfile ADD directive) the whole thing bloats up to about 1.5GB(!). Not ideal as you can immagine.

I would like to have this '-v /distrib/ready2installApp:/distrib' directive still possible (as it is today in the Dockerfile)
but
I would like to disassociate the declarative build process (infrastructure as code) from the container run-time deployable artifact. I do not want to have to deal with the dead weight of 1.5GB that I do not need.
--

Could we have an --unmount-volume option that I can run at the end of the Dockerfile?
or
Given how Volume works right now in a Dockerfile, maybe we need a new Dockerfile directive for a temporary volume that people use while installing? I think the Puppet example supplied by @fatherlinux was on a similar line...
or
Whatever you guys can think of.
The objective is avoiding to have to carry around all that dead weight that is useless for a deployed App or Service. However that dead weight is necessary @install-time. Not Everybody has a simple "yum install" from the official repositories. :)

thank you very much

### PR DESCRIPTION
**Pull request by rdsubhas**
#### What this does

- Allows `--mount`s to be specified during `docker build`
- Example: `docker build --mount type=bind,source=/hostpath,target=/containerpath ...`
- Various different use cases have been documented in #14080 

#### Checklist

- [ ] Fixes #14080 (eventually)
- [x] Initial barebones commit, adds and passes through mount flags in client & server
- [x] Dockerfile is able to `ls` the mount only while building, and does not exist in image metadata or when running
- [x] Use `--mount` instead of `-v`
- [x] Existing test suite passes, no breakage in any existing behavior
- [ ] Design review and feedback, `/build` api, cli, etc
- [ ] TBD: Add test cases
- [ ] TBD: Add build cache invalidation (equivalent to `--no-cache`) if mount definitions are changed
- [ ] TBD: Add docs, `mount` options are less documented in general, difference between `docker run -v` and `docker build --mount`, general warnings and disclaimers, how to invalidate cache, etc
- [ ] TBD: Edge cases, file permissions when copying from volume, post-build leftovers (ideally none except an empty folder is present), running inside a container with docker.sock bind-mounted, volume drivers, mounts overlapping with `ADD/COPY` inside build context/workdir, etc

Signed-off-by: Subhas Dandapani <rdsubhas@gmail.com>

Obligatory picture of how it feels like in the beginning. Slowly warming up. [Great developer docs](https://docs.docker.com/opensource/project/who-written-for/) 👍 It was really shocking how simple this was to get working though. All these years of orchestrating `docker build` as much as (or sometimes more than) `docker run` for sake of performance and developer productivity feels like a lie. Hope we can move forward...

that works?

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @tonistiigi @cpuguy83 PTAL

**Comment by rdsubhas** [comment]
@thaJeztah this is not ready yet, I opened only for running the specs.

Also I believe I misunderstood @cpuguy83's comments. Will change `volume` to `mounts`

**Comment by AkihiroSuda** [comment]
also, consideration for build cache is needed

**Comment by rdsubhas** [comment]
@thaJeztah this is now ready for design review please, before I go deep into covering test cases, I'd like to make sure I'm on the right path 👍

**Comment by thaJeztah** [comment]
Thanks! I added this for the upcoming maintainers meeting, but perhaps @tonistiigi has time to look before that. He's working on some other improvements to the builder, so we should verify this addition doesn't "conflict" with the other enhancements

**Comment by pyros2097** [comment]
Don't know if this is useful. But it seems these guys have already solved it https://github.com/grammarly/rocker.

**Comment by alexellis** [comment]
Rocker could be a good option for build-time mounts. Have you had a chance to check it out @rdsubhas ?

**Comment by tonistiigi** [comment]
Hi @rdsubhas,

As you have probably already read, this has been discussed many times previously. There are many reasons why I would feel hesitant about adding such a feature. It makes builds not self-contained, encourages creating broken dockerfiles, builds can't be tracked to immutable sources, invalidates caching, breaks remote usage of Docker(as bind mounts are not part of API) etc.

Let's look at the actual use-cases behind this solutions:

#### Performance, mounting is faster than copying context when there are lots of files

I'm working on a prototype of incremental contexts sends and can share more details soon. This means that repeated builds should not perform worse when the size of the context increases.

Once this is in place and we have an immutable reference to the input data in daemon, we could think about mounting it to commands read-only if that makes more sense than `ADD/COPY` in some cases. This wouldn't create any of the problems raised earlier.

There are many other cases how the build performance could be improved. If you are interested in helping out in some of them we should discuss more.

In an earlier issue, there was also a comment about this being a workaround for smaller images. I've proposed https://github.com/docker/docker/pull/31257 for that.

#### Need to share secret data to a running process

There is an open PR #30637. The benefit of that PR instead of this one is that it is clearly limiting the use case to secret data what makes it harder to have a completely broken Dockerfile or misuse it in another way. Secrets are clearly defined and we could combine them with a Dockerfile directive in the future so that required secrets could be defined in a similar way `ARG` are defined atm. With this method, secrets are also immutable.

Another way to fix the secret problem would be to use an external service that provides trusted information (from swarm cluster, client, etc).

#### Need to include directory outside of build context

While this is also encouraging broken Dockerfiles, sometimes, especially while developing and overriding other projects, it seems helpful to not require everything to be in the same folder. A solution for that has been discussed in https://github.com/docker/docker/issues/30101#issuecomment-279901542 . It is quite similar to this PR but "mounts" the data to context instead of the container.

**Comment by tianon** [comment]
The main use case for something like this that sticks out to me (and was mentioned in #14080 but not explicitly here) is the case of Gentoo-based images.

For Gentoo, the "packaging" files are stored in `/usr/portage`, take up a very large amount of space, change often, and mirrors are unfriendly to (and will sometimes actively ban) folks who update too often.  Being able to mount that directory from the host or from another image would be really useful for making images based on Gentoo actually feasible (and it's prohibitively large, so putting it in the context somehow is going to be painful too -- on my current host, `/usr/portage` is ~21GB, but if I exclude `distfiles` it's ~735MB).

Relatedly, while building packages, the source of the packages being built gets downloaded to `/usr/portage/distfiles` by default.

So in summary, the main use cases I see for "outside content" (based mostly on @tonistiigi's list above) are:

1. cache data (gem/npm/pip cache files, Gentoo's `distfiles`, etc)

2. secrets (SSH keys for fetching privileged code, etc)

3. extra directories outside the build context (which I'd put into the following two sub-use-cases, which is the main divergence from @tonistiigi's list)

   1. more files needed by the build, such as shared common files (could be transmitted via context somehow)

   2. things like Gentoo's `/usr/portage`, which shouldn't necessarily be included in the final image, but are needed while building (and can be prohibitively large)

**Comment by pyros2097** [comment]
@tonistiigi I agree that making docker build mutable and inconsistent is not right but now one in their right might would do such a build for a production image. In our projects we have a ton of node_modules and each time someone in the team changes one of our private packages we need to do a clean npm install again in our development docker container and that takes almost 30 minutes and we also get rate limited by npm many times for doing too many updates in a day.  I tried all sorts of ways to cache the node_modules and even tried yarn to save the packages offline so that they will be available in the build context but to no avail.  Even tried rocker and that also didn't work.

The only other way I see of solving this is running the npm install command in our existing container to update the packages and commit the container as the new latest image and that would solve our developer woes. So its okay if its not immutable since its development but can we have a command that does this or I need to write a python script to do this from now on.

Docker has eased our devops work but substantially increased the developers work and we get pounded by the developers all the time because of this.

**Comment by rdsubhas** [comment]
Sorry for responding late, different timezones,

@alexellis Yes indeed we have checked out rocker, but wherever possible we'd like to be compliant with normal upstream docker, especially since its moved to shorter release cycles.

@tonistiigi I think @tianon and others in #14080 have really collated a long list of issues. Some of us work in the corporate world where we don't really have the simplicity of doing short and sweet `npm install/run` or simple `mvn package and java -jar`. Some of us work with huge distributions. Some on low bandwidth internet connections, or low powered laptops and so on. There is a wealth of problems out here, and while we all appreciate idempotent builds, it doesn't seem to be a reality. Even docker builds without this are not really idempotent. Incremental build contexts is really nice, but again its pushing the problem down the line instead of solving it (e.g. HTTP caching issues, etc).

What we see is simple: "a `Dockerfile` builds image from sources". All the hacks, like a build orchestrator to do multi-container docker builds, rocker, docker-compose orchestration, etc - have all **already broken** the Dockerfile - i.e. one just can't do `docker build` anymore. So what gives...

**Comment by thaJeztah** [comment]
There's a proposal for `--mount` and secrets  in multi-stage build; for those following this issue, please participate in the discussions there; https://github.com/moby/moby/issues/32507 https://github.com/moby/moby/issues/33343

**Comment by rdsubhas** [comment]
@thaJeztah thanks! will close this...

**Comment by thaJeztah** [comment]
@rdsubhas thanks! I didn't want to close because no decision was made yet, but would definitely appreciate your input on those proposals 👍
